protocols/cosi/cosi.go

```diff
@@ -0,0 +1,338 @@
+func (c *CoSi) handleAnnouncement(in *Announcement) error {
+	log.Lvl3("Message:", c.Message)
+	// If we have a hook on announcement call the hook
+	if c.announcementHook != nil {
+		return c.announcementHook()
+	}
+
+	// If we are leaf, we should go to commitment
+	if c.IsLeaf() {
+		return c.handleCommitment(nil)
+	}
+	// send to children
+	return c.SendToChildren(in)
+}
+
+// handleAllCommitment relay the commitments up in the tree
+// It expects *in* to be the full set of messages from the children.
+// The children's commitment must remain constants.
+func (c *CoSi) handleCommitment(in *Commitment) error {
+	if !c.IsLeaf() {
+		// add to temporary
+		c.tempCommitLock.Lock()
+		c.tempCommitment = append(c.tempCommitment, in.Comm)
+		c.tempCommitLock.Unlock()
+		// do we have enough ?
+		// TODO: exception mechanism will be put into another protocol
+		if len(c.tempCommitment) < len(c.Children()) {
+			return nil
+		}
+	}
+	log.Lvl3(c.Name(), "aggregated")
+	// pass it to the hook
+	if c.commitmentHook != nil {
+		return c.commitmentHook(c.tempCommitment)
+	}
+
+	// go to Commit()
+	out := c.cosi.Commit(nil, c.tempCommitment)
+
+	// if we are the root, we need to start the Challenge
+	if c.IsRoot() {
+		return c.startChallenge()
+	}
+
+	// otherwise send it to parent
+	outMsg := &Commitment{
+		Comm: out,
+	}
+	return c.SendTo(c.Parent(), outMsg)
+}
+
+// StartChallenge starts the challenge phase. Typically called by the Root ;)
+func (c *CoSi) startChallenge() error {
+	challenge, err := c.cosi.CreateChallenge(c.Message)
+	if err != nil {
+		return err
+	}
+	out := &Challenge{
+		Chall: challenge,
+	}
+	log.Lvl3(c.Name(), "Starting Chal=", fmt.Sprintf("%+v", challenge), " (message =", string(c.Message))
+	return c.handleChallenge(out)
+
+}
+
+// handleChallenge dispatch the challenge to the round and then dispatch the
+// results down the tree.
+func (c *CoSi) handleChallenge(in *Challenge) error {
+	log.Lvl3(c.Name(), "chal=", fmt.Sprintf("%+v", in.Chall))
+	c.cosi.Challenge(in.Chall)
+
+	if c.challengeHook != nil {
+		c.challengeHook(in.Chall)
+	}
+
+	// if we are leaf, then go to response
+	if c.IsLeaf() {
+		return c.handleResponse(nil)
+	}
+
+	// otherwise send it to children
+	return c.SendToChildren(in)
+}
+
+// handleResponse brings up the response of each node in the tree to the root.
+func (c *CoSi) handleResponse(in *Response) error {
+	if !c.IsLeaf() {
+		// add to temporary
+		c.tempResponseLock.Lock()
+		c.tempResponse = append(c.tempResponse, in.Resp)
+		c.tempResponseLock.Unlock()
+		// do we have enough ?
+		log.Lvl3(c.Name(), "has", len(c.tempResponse), "responses")
+		if len(c.tempResponse) < len(c.Children()) {
+			return nil
+		}
+	}
+
+	defer func() {
+		// protocol is finished
+		close(c.done)
+		c.Done()
+	}()
+
+	log.Lvl3(c.Name(), "aggregated")
+	outResponse, err := c.cosi.Response(c.tempResponse)
+	if err != nil {
+		return err
+	}
+
+	if c.responseHook != nil {
```

Comment:
@ineiti could you push your changes that you did in #664 (snp17) here so it's compatible with randhound ?